src/Evolution/Systems/GrMhd/GhValenciaDivClean/Characteristics.cpp

```diff
@@ -0,0 +1,141 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/GrMhd/GhValenciaDivClean/Characteristics.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/Characteristics.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.hpp"
+#include "Evolution/Systems/RelativisticEuler/Valencia/Characteristics.hpp"
+#include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"  // IWYU pragma: keep
+#include "PointwiseFunctions/Hydro/Tags.hpp"              // IWYU pragma: keep
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+// IWYU pragma: no_forward_declare EquationsOfState::EquationOfState
+// IWYU pragma: no_forward_declare Tensor
+
+namespace grmhd::GhValenciaDivClean {
+template <size_t ThermodynamicDim>
+void characteristic_speeds(
+    const gsl::not_null<std::array<DataVector, 13>*> char_speeds,
+    const Scalar<DataVector>& rest_mass_density,
+    const Scalar<DataVector>& specific_internal_energy,
+    const Scalar<DataVector>& specific_enthalpy,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& spatial_velocity,
+    const Scalar<DataVector>& lorentz_factor,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& magnetic_field,
+    const Scalar<DataVector>& lapse, const tnsr::I<DataVector, 3>& shift,
+    const tnsr::ii<DataVector, 3, Frame::Inertial>& spatial_metric,
+    const tnsr::i<DataVector, 3>& unit_normal,
+    const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
+        equation_of_state,
+    const Scalar<DataVector>& gh_constraint_gamma1) noexcept {
+  const size_t num_grid_points = get(lapse).size();
+  auto& achar_speeds = *char_speeds;
+  for (size_t idx = 0; idx < 13; ++idx) {
+    if (achar_speeds.at(idx).size() != num_grid_points) {
+      achar_speeds.at(idx) = DataVector(num_grid_points);
+    }
+  }
+
+  const auto char_speeds_valencia =
+      grmhd::ValenciaDivClean::characteristic_speeds(
```

Comment:
You are currently not gaining anything by having a not_null version of this function, because you allocate by using the return-by-value versions from the component systems.  To avoid this, you'll need to move the passed-in DataVectors from the array<13> into an array<9> and an array<4>, call the component systems' functions on those, and then move the DataVectors back.